node/src/Cardano/Node/Client.hs

```diff
@@ -0,0 +1,136 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE LambdaCase    #-}
+
+module Cardano.Node.Client
+    ( -- Node Client
+      NodeClient
+    , ClientError(..)
+    , fromServantError
+
+    -- * HTTP instance
+    , mkHttpClient
+    ) where
+
+import           Universum
+
+import           Data.Aeson (FromJSON)
+import qualified Data.Aeson as Aeson
+import           Network.HTTP.Client (Manager)
+import           Network.HTTP.Media.MediaType (MediaType)
+import           Servant ((:<|>) (..))
+import           Servant.Client (BaseUrl (..), ClientEnv (..), ClientM,
+                     GenResponse (..), Response, ServantError, client,
+                     runClientM)
+import qualified Servant.Client as Servant
+
+import           Pos.Chain.Txp (Utxo)
+import           Pos.Node.API (API, ForceNtpCheck, NodeInfo, NodeSettings)
+import           Pos.Util.Jsend (ResponseStatus (..))
+import           Pos.Util.Servant (APIResponse (..))
+import qualified Pos.Web.Api as Legacy
+import           Pos.Web.Types (CConfirmedProposalState)
+
+
+-- * Node Client
+
+data NodeClient m
+    = NodeClient
+    { getUtxo
+        :: m Utxo
+
+    , getConfirmedProposals
+        :: m [CConfirmedProposalState]
+
+    , getNodeSettings
+        :: m NodeSettings
+
+    , getNodeInfo
+        :: ForceNtpCheck
+        -> m NodeInfo
+
+    , applyUpdate
+        :: m ()
+
+    , postponeUpdate
+        :: m ()
+    } deriving (Generic)
+
+
+data ClientError a
+    = KnownError a
+    | DecodeFailure Text Response
+    | UnsupportedContentType MediaType Response
+    | InvalidContentTypeHeader Response
+    | ConnectionError Text
+    deriving (Show, Generic, Eq)
+
+instance Exception a => Exception (ClientError a) where
+    toException = \case
+        KnownError a ->
+            toException a
+        DecodeFailure t r ->
+            toException (Servant.DecodeFailure t r)
+        UnsupportedContentType m r ->
+            toException (Servant.UnsupportedContentType m r)
+        InvalidContentTypeHeader r ->
+            toException (Servant.InvalidContentTypeHeader r)
+        ConnectionError t ->
+            toException (Servant.ConnectionError t)
```

Comment:
this instance is a little odd to me -- IIRC it is going ot make it impossible to catch `ClientError` as an exception! The definition of `throwIO` is:

```haskell
throwIO :: Exception e => e -> IO a
throwIO e = IO (raiseIO# (toException e))
```

So, if we have a value like `err = ConnectionError "hello" :: ClientError OtherExceptionType`, and we do `throwIO err`, then it will "convert" it into a `ServantError`. If we later try to `catch` a `ClientError`, then it's going to fail, because we actually threw a `ServantError`. Same for the `KnownError` case.

It should be fine to write 

```haskell
instance (Exception a) => Exception (ClientError a)
```

Reply:
I checked indeed. Very dodgy :upside_down_face:  ...